mantidproject/mantid
# Add option to exclude/annotate runs in the search results

**Report to:** Max at ISIS

This is still work in progress

This PR adds functionality to allow users to exclude runs from an experiment and to annotate a reason why. It also allows them to add a general comment to runs for their own notes. The main changes are:
- Add columns to the search results table for the exclude reason and comment. Runs are excluded from transfer to the main table if the exclude reason is set and the row will be highlighted blue to indicate it will be excluded (similar to invalid runs).
- Include search results rows in the encoder/decoder so that this information is included when saving the batch to a file.
- Ensure search results are not cleared when re-running a search with the same criteria (ensures the annotations are not lost)
- Warn the user if changes will be lost e.g. if running a search with different criteria or changing the instrument

There has also been quite a lot of tidying in the tests to make adding the new tests clearer, and to fix some problems with existing tests not being run.

**To test:**

Should be tested by someone at ISIS

- Open the ISIS Reflectometry interface
- Set the instrument to INTER and enter investigation id `1120015` and cycle `11_3` and click Search
- The search results list should be populated. The description column should be expanded but still truncated to fit the Exclude and Comment columns in.
- Try entering text in the Exclude and Comment columns for some rows. If Exclude is set the row should turn blue to indicate it will be excluded. Try clearing it and it should return to white.
- Try selecting some of the rows you edited and click Transfer. White rows should be copied to the main table but blue rows should not.
- Try clicking Search again and check that your edits are not cleared.
- Delete the rows in the main table. Click Autoprocess and check that your changes are not cleared. Again, the white rows should be copied to the main table and blue rows should not.
- Click Pause. Clear an Exclude reason and click Autoprocess. The run should now be added to the main table. (Note that this only works when including new runs; if you exclude a run after it has already been added to the main table, it will not be removed)
- Check you get a warning about losing changes when you:
  - close the interface
  - change the instrument
  - change the search criteria and re-do the search or autoprocess
- Go to Batch->Save and save your changes to a file. Check that the warnings are gone when e.g. on changing instrument.
- Go to Batch->Load and load your file. Check that your changes are re-instated.

Fixes #29148

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
